Sources/ComposableArchitecture/Instrumentation.swift

```diff
@@ -0,0 +1,142 @@
+///   3. The ``Action`` is passed to ``Store.send``
+///   4. The ``Instrumentation.Store.willSend`` callback is called
+///   5. The ``Instrumentation.Store.willProcessEvents`` callback is called
+///   6. The ``Action`` is passed to the ``Store.reducer`` and acted upon
+///   7. The ``Store.reducer`` returns an ``Effect.none`` producing no further ``Action``s
+///   8. The ``Instrumentation.Store.didProcessEvents`` callback is called
+///   9. The ``Instrumentation.Store.willScope`` callback is called
+///   10. The ``Store.state.value`` is updated
+///   11. The ``ViewStore`` is updated with the ``Store.state``s new value
+///   12. The ``Instrumentation.ViewStore.willDeduplicate`` callback is called
+///   13. The state is deduplicated
+///   14. The ``Instrumentation.ViewStore.didDeduplicate`` callback is called
+///   15. The ``Instrumentation.ViewStore.stateWillChange`` callback is called
+///   16. The ``ViewStore.state`` value is updated
+///   17. All subscribers to the ``ViewStore.state`` are updated with the new state value
+///   18. The ``Instrumentation.ViewStore.stateDidChange`` callback is called
+///   19. The ``Instrumentation.Store.didScope`` callback is called
+///   20. The ``Instrumentation.Store.didSend`` callback is called
+///   21. The ``Instrumentation.ViewStore.didSend`` callback is called
+public class Instrumentation {
+  public typealias Trigger = (EventInfo) -> Void
+
+  /// Container for the information that will be provided to tracking/instrumentation implementations.
+  public struct EventInfo: CustomStringConvertible {
+    internal init(type: String, action: String = "", tags: [String: String] = [:]) {
+      self.type = type
+      self.action = action
+      self.tags = tags
+    }
+
+    /// The Swift type of object that is operating. This will generally be of the type `Store<State, Action>` or
+    /// `ViewStore<State, Action>` with the `State` and `Action` types properly filled in. With this information it
+    /// _should_ be possible to identify which ``Store`` or ``ViewStore`` (or other type) is operating.
+    public let type: String
+
+    /// A ``String`` generated for the ``Action`` that was sent, when available. There are operations that may not have
+    /// an ``Action`` available and so this may be an empty string. The value is generated using ``String(describing:)``
+    /// and so is dependent on the Swift runtime metadata. If that metadata is removed in some way then this value will
+    /// be empty most likely.
+    public var action: String
+
+    /// A dictionary of tags that the library thinks are valuable to include in the tracking/instrumentation. There is
+    /// obviously no requirement to use these, but they are there just in case.
+    public var tags: [String: String]
+
+    public var description: String {
+      guard !action.isEmpty else {
+        return "\(type)"
+      }
+
+      return "\(type): \(action)"
```

Comment:
FWIW l.59-63 can be rewritten as just:
`type + (action.isEmpty ? "" : " "  + action)`

Reply:
That string concatenation is not free though. :-/

Reply:
Of course I wrote that, and now look back and realize I was being naive. I have String interpolation of ... another String. So that isn't gonna be any free-er than String concatenation probably.

Will update to recommendation.